File: pipeline/kite9-pipeline-common/src/main/java/com/kite9/k9server/command/xml/replace/ReplaceXML.java
package com.kite9.k9server.command.xml.replace;

import org.kite9.diagram.dom.elements.ADLDocument;
import org.springframework.http.HttpStatus;
import org.w3c.dom.Element;
import org.w3c.dom.Node;

import com.kite9.k9server.adl.holder.pipeline.ADLDom;
import com.kite9.k9server.command.CommandException;

public class ReplaceXML extends AbstractReplaceCommand<Element, Element> {
	
	@Override
	protected Element getFromContent(ADLDom adl) {
		Element out = decodeElement(from, adl);
		if (!out.getAttribute("id").equals(fragmentId)) {
			throw new CommandException(HttpStatus.BAD_REQUEST, "ReplaceXML should preserve ID");
		}
		return out;
	}

	@Override
	protected Element getToContent(ADLDom adl) {
		Element out = decodeElement(to, adl);
		if (!out.getAttribute("id").equals(fragmentId)) {
			throw new CommandException(HttpStatus.BAD_REQUEST, "ReplaceXML should preserve ID");
		}
		return out;
	}

	@Override
	protected Element getExistingContent(ADLDom in) {
		return findFragmentElement(in.getDocument(), fragmentId);
	}

	@Override
	protected void doReplace(ADLDom in, Element old, Element toContent, Element fromContent) {
		ADLDocument doc = in.getDocument();
		doc.adoptNode(toContent);
		Node into = old.getParentNode();
		into.replaceChild(toContent, old);
		ensureParentElements(into, toContent);
		
		LOG.info("Processed replace XML of "+fragmentId);
	}

	@Override
	protected Mismatch same(Element existing, Element with) {
		return twoElementsAreIdentical(existing, with);
	}

	

}

File: pipeline/kite9-pipeline-common/src/main/kotlin/com/kite9/pipeline/command/xml/replace/ReplaceXML.kt
package com.kite9.pipeline.command.xml.replace

import com.kite9.pipeline.adl.holder.pipeline.ADLDom
import com.kite9.pipeline.command.Command
import com.kite9.pipeline.command.CommandContext
import com.kite9.pipeline.command.CommandException
import org.w3c.dom.Document
import org.w3c.dom.Element

class ReplaceXML : AbstractReplaceCommand<Element, Element>() {

    override fun getFromContent(context: ADLDom, ctx: CommandContext): Element {
        val out: Element = ctx.decodeElement(from, context)
        if (out.getAttribute("id") != fragmentId) {
            throw CommandException(BAD_REQUEST, "ReplaceXML should preserve ID", this, null)
        }
        return out
    }

    override fun getToContent(context: ADLDom, ctx: CommandContext): Element {
        val out: Element = ctx.decodeElement(to, context)
        if (out.getAttribute("id") != fragmentId) {
            throw CommandException(BAD_REQUEST, "ReplaceXML should preserve ID", this, null)
        }
        return out
    }

    override fun getExistingContent(o: ADLDom, ctx: CommandContext): Element {
        return findFragmentElement(o.document, fragmentId, ctx)
    }

    override fun doReplace(
        on: ADLDom,
        site: Element,
        toContent: Element,
        fromContent: Element,
        ctx: CommandContext
    ) {
        val doc: Document = on.document
        doc.adoptNode(toContent)
        val into = site.parentNode
        into.replaceChild(toContent, site)
        ensureParentElements(into, toContent, ctx)
        ctx.log("Processed replace XML of $fragmentId")
    }

    private fun same(existing: Element, with: Element, ctx: CommandContext): Command.Mismatch? {
        return ctx.twoElementsAreIdentical(existing, with)
    }
}